File: solutions/go/ThirdMaximumNumber.go
import "math"

func thirdMax(nums []int) int {
	var a, b, c int

	a, b, c = math.MinInt64, math.MinInt64, math.MinInt64

	for _, v := range nums {
		if v <= a || v == b || v == c {
			continue
		}

		if v > c {
			a, b, c = b, c, v
		} else if v > b {
			a, b = b, v
		} else {
			a = v
		}
	}

	if a == math.MinInt64 {
		return c
	}

	return a
}

File: solutions/cpp/ThirdMaximumNumber.cc
class Solution {
public:
    int thirdMax(vector<int>& nums) {
        long long a, b, c;

        a = b = c = LLONG_MIN;

        for (int v : nums) {
            if (v <= a || v == b || v == c) {
                continue;
            }

            if (v > c) {
                a = b;
                b = c;
                c = v;
            } else if (v > b) {
                a = b;
                b = v;
            } else {
                a = v;
            }
        }

        return a == LLONG_MIN ? c : a;
    }
};
